src/axom/spin/BVH.hpp

```diff
@@ -0,0 +1,848 @@
+
+  if ( point[0]  > s1[3] ) in_left = false;
+  if ( point[1]  > s2[0] ) in_left = false;
+  if ( point[2]  > s2[1] ) in_left = false;
+
+  return in_left;
+}
+
+//------------------------------------------------------------------------------
+template< typename FloatType >
+AXOM_HOST_DEVICE
+bool InLeft( const internal::linear_bvh::Vec< FloatType, 2>& point,
+             const internal::linear_bvh::Vec< FloatType, 4>& s1,
+             const internal::linear_bvh::Vec< FloatType, 4>& s2 )
+{
+  bool in_left = true;
+  if ( point[0]  < s1[0] ) in_left = false;
+  if ( point[1]  < s1[1] ) in_left = false;
+
+  if ( point[0]  > s1[3] ) in_left = false;
+  if ( point[1]  > s2[0] ) in_left = false;
+
+  return in_left;
+}
+
+//------------------------------------------------------------------------------
+template< typename FloatType >
+AXOM_HOST_DEVICE
+bool InRight( const internal::linear_bvh::Vec< FloatType, 3>& point,
+              const internal::linear_bvh::Vec< FloatType, 4>& s2,
+              const internal::linear_bvh::Vec< FloatType, 4>& s3 )
+{
+  bool in_right = true;
+
+  if ( point[0] < s2[2] ) in_right = false;
+  if ( point[1] < s2[3] ) in_right = false;
+  if ( point[2] < s3[0] ) in_right = false;
+
+  if ( point[0] > s3[1] ) in_right = false;
+  if ( point[1] > s3[2] ) in_right = false;
+  if ( point[2] > s3[3] ) in_right = false;
+
+  return in_right;
+}
+
+//------------------------------------------------------------------------------
+template< typename FloatType >
+AXOM_HOST_DEVICE
+bool InRight( const internal::linear_bvh::Vec< FloatType, 2>& point,
+              const internal::linear_bvh::Vec< FloatType, 4>& s2,
+              const internal::linear_bvh::Vec< FloatType, 4>& s3 )
+{
+  bool in_right = true;
+
+  if ( point[0] < s2[2] ) in_right = false;
+  if ( point[1] < s2[3] ) in_right = false;
+
+  if ( point[0] > s3[1] ) in_right = false;
+  if ( point[1] > s3[2] ) in_right = false;
+
+  return in_right;
+}
+
+}
+
+//------------------------------------------------------------------------------
+//  BVH IMPLEMENTATION
+//------------------------------------------------------------------------------
+template< int NDIMS, typename ExecSpace, typename FloatType >
+BVH< NDIMS, ExecSpace, FloatType >::BVH( const FloatType* boxes,
+                                         IndexType numItems ) :
+  m_numItems( numItems ),
+  m_boxes( boxes )
+{
+
+}
+
+//------------------------------------------------------------------------------
+template< int NDIMS, typename ExecSpace, typename FloatType >
+BVH< NDIMS, ExecSpace, FloatType >::~BVH()
+{
+  m_bvh.deallocate();
+}
+
+//------------------------------------------------------------------------------
+template< int NDIMS, typename ExecSpace, typename FloatType >
+int BVH< NDIMS, ExecSpace, FloatType >::build()
+{
+  // STEP 0: set the default memory allocator to use for the execution space.
+  umpire::Allocator current_allocator = axom::getDefaultAllocator();
```

Comment:
You could remove the dependency on umpire here by passing an Allocator ID (int) to the `linear_bvh` methods.